gapii/client/capture.go

```diff
@@ -181,41 +186,109 @@ func (p *Process) Capture(ctx context.Context, start task.Signal, stop task.Sign
 	conn := p.conn
 	defer conn.Close()
 
-	var count siSize
-	started := false
+	if (p.Options.Flags & DeferStart) != 0 {
+		go func() {
+			if start.Wait(ctx) {
+				conn.Write([]byte{0x1, messageStartTrace})
+			}
+		}()
+	}
+
+	go func() {
+		if stop.Wait(ctx) {
+			conn.Write([]byte{0x1, messageEndTrace})
+		}
+	}()
+
+	buf := make([]byte, 64*1024)
+	var bufSize, bufPos int
+	var headerPos uint
+	var chunkSize uint64
+	var readErr error
+	var count, remain siSize
+	var messageType byte = messageInvalid
+	var bufErr []byte
+mainLoop:
 	for {
-		if task.Stopped(ctx) || stop.Fired() {
+		if task.Stopped(ctx) {
 			log.I(ctx, "Stop: %v", count)
-			break
+			break mainLoop
 		}
```

Comment:
Instead of reading into a buffer, then tearing apart that buffer piece by piece, I would recommend something a bit different.
We have io.ReadFull() which will read EXACTLY the number of bytes you want, (or return an error), We can turn this into a simpler loop of the form:

```
headerBuffer := make([]byte, 9)
if _, err := io.ReadFull(conn, headerBuffer); err != nil {
    return nil, err
}
// headerBuffer now definitely contains the header
size := parseHeader(headerBuffer).size
messageBuffer := make([]byte, size)
if _, err := io.ReadFull(conn, messageBuffer); err != nil {
    return nil, err
}
messageBuffer now contains the entire message
```

Reply:
I would also suggest having a parseHeader method sitting beside the description of the header something like:
```
func ParseHeader(data []byte) (Header, error) {
    if len(data) != 9 {
     return fmt.Errorf("Invalid header length"); 
   }
   // parse all 9 bytes of the header at once into a header struct
   return header, nil
}
```

Reply:
Yep, I thought read small chunks while using CopyN, ReadFull, etc. would be bad for the garbage collector and socket synchronization. Pascal convinced me of the opposite, so I already revised it, see latest commit.